api/gateway-operator/v2alpha1/controlplane_types.go

```diff
@@ -0,0 +1,236 @@
+	//
+	// +optional
+	// +kubebuilder:default={type: all}
+	WatchNamespaces *operatorv1beta1.WatchNamespaces `json:"watchNamespaces,omitempty"`
+
+	// FeatureGates is a list of feature gates that are enabled for this ControlPlane.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=32
+	FeatureGates []ControlPlaneFeatureGate `json:"featureGates,omitempty"`
+
+	// Controllers defines the controllers that are enabled for this ControlPlane.
+	//
+	// +optional
+	// +listType=map
+	// +listMapKey=name
+	// +kubebuilder:validation:MaxItems=32
+	Controllers []ControlPlaneController `json:"controllers,omitempty"`
+
+	// AdminAPI defines the configuration for the Kong Admin API.
+	//
+	// +optional
+	AdminAPI *ControlPlaneAdminAPI `json:"adminAPI,omitempty"`
+}
+
+// ControlPlaneDataPlaneTarget defines the target for the DataPlane that the ControlPlane
+// is responsible for configuring.
+//
+// +kubebuilder:validation:XValidation:message="URL has to be provided when type is set to url",rule="self.type != 'url' || has(self.url)"
+// +kubebuilder:validation:XValidation:message="Name cannot be provided when type is set to url",rule="self.type != 'url' || !has(self.name)"
+// +kubebuilder:validation:XValidation:message="Name has to be provided when type is set to name",rule="self.type != 'name' || has(self.name)"
+// +kubebuilder:validation:XValidation:message="URL cannot be provided when type is set to name",rule="self.type != 'name' || !has(self.url)"
+type ControlPlaneDataPlaneTarget struct {
+	// Type indicates the type of the DataPlane target.
+	//
+	// +kubebuilder:validation:Enum=url;name
+	// +kubebuilder:validation:Required
+	Type string `json:"type,omitempty"`
+
+	// URL is the URL of the DataPlane target. This is used for configuring
+	// externally managed DataPlanes like those installed independently with Helm.
+	//
+	// +optional
+	URL string `json:"url,omitempty"`
+
+	// Name is the name of the DataPlane to configure.
+	//
+	// +optional
+	Name string `json:"name,omitempty"`
+}
+
+// ControlPlaneAdminAPI defines the configuration for the Kong Admin API that
+// a ControlPlane when configuring the DataPlane.
+type ControlPlaneAdminAPI struct {
+	// Workspace indicates the Kong Workspace to use for the ControlPlane.
+	// If left empty then no Kong workspace will be used.
+	//
+	// +optional
+	Workspace string `json:"workspace,omitempty"`
+}
+
+// ControlPlaneController defines a controller state for the ControlPlane.
+// It overrides the default behavior as defined in the deployed operator version.
+//
+// +apireference:kgo:include
+type ControlPlaneController struct {
+	// Name is the name of the controller.
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name,omitempty"`
+
+	// Enabled indicates whether the controller is enabled or not.
+	//
+	// +kubebuilder:validation:Required
+	Enabled *bool `json:"enabled,omitempty"`
+}
```

Comment:
I don't think this is great UX: let's say a user is running a CP with X and Y controllers explicitly enabled. How this struct works does not define the current state of the world: as a user, I should go to the KGO docs, see which controllers are enabled by default, and cross that data with the one here defined. OTOH, I understand the need to have such an implicit default. We could add some information in the status to list all the controllers actually enabled for the specific instance, to give users a good feedback on what's going on. WDYT?

Reply:
I agree. This is a good idea to have this in status. Let me work on that